core/embed/trezorhal/stm32f4/display/st-7789/display_driver.c

```diff
@@ -0,0 +1,222 @@
+  display_io_init_fmc();
+  display_panel_init();
+  display_panel_set_little_endian();
+  backlight_pwm_init();
+
+#ifdef XFRAMEBUFFER
+  display_io_init_te_interrupt();
+#endif
+}
+
+void display_reinit(void) {
+  display_driver_t* drv = &g_display_driver;
+  memset(drv, 0, sizeof(display_driver_t));
+
+  // Reinitialize FMC to set correct timing
+  // We have to do this in reinit because boardloader is fixed.
+  display_io_init_fmc();
+
+  // Important for model T as this is not set in boardloader
+  display_panel_set_little_endian();
+  display_panel_init_gamma();
+  backlight_pwm_reinit();
+
+#ifdef XFRAMEBUFFER
+  display_io_init_te_interrupt();
+#endif
+}
+
+void display_finish_actions(void) {
+#ifdef XFRAMEBUFFER
+#ifndef BOARDLOADER
+  bg_copy_wait();
+#endif
+#endif
+}
+
+int display_set_backlight(int level) {
+#ifdef XFRAMEBUFFER
+#ifndef BOARDLOADER
+  // wait for DMA transfer to finish before changing backlight
+  // so that we know that panel has current data
+  if (backlight_pwm_get() != level && !is_mode_handler()) {
+    bg_copy_wait();
+  }
+#endif
+#endif
+
+  return backlight_pwm_set(level);
+}
+
+int display_get_backlight(void) { return backlight_pwm_get(); }
+
+int display_set_orientation(int angle) {
+  display_driver_t* drv = &g_display_driver;
+
+  if (angle != drv->orientation_angle) {
+    if (angle == 0 || angle == 90 || angle == 180 || angle == 270) {
+      drv->orientation_angle = angle;
+
+#ifdef XFRAMEBUFFER
+      memset(physical_frame_buffer_0, 0, sizeof(physical_frame_buffer_0));
+      memset(physical_frame_buffer_1, 0, sizeof(physical_frame_buffer_1));
+#endif
+
+      display_panel_set_window(0, 0, DISPLAY_RESX - 1, DISPLAY_RESY - 1);
+      for (uint32_t i = 0; i < DISPLAY_RESX * DISPLAY_RESY; i++) {
+        // 2 bytes per pixel because we're using RGB 5-6-5 format
+        ISSUE_PIXEL_DATA(0x0000);
+      }
+
+      display_panel_rotate(angle);
+    }
+  }
+
+  return drv->orientation_angle;
+}
+
+int display_get_orientation(void) {
+  display_driver_t* drv = &g_display_driver;
+
+  return drv->orientation_angle;
+}
+
+#ifndef XFRAMEBUFFER
+void display_refresh(void) {
+  // if the framebuffer is not used the implementation is empty
+}
+#endif
+
+void display_wait_for_sync(void) {
+#ifdef DISPLAY_TE_PIN
+  uint32_t id = display_panel_identify();
+  if (id && (id != DISPLAY_ID_GC9307)) {
+    // synchronize with the panel synchronization signal
+    // in order to avoid visual tearing effects
+    while (GPIO_PIN_SET == HAL_GPIO_ReadPin(DISPLAY_TE_PORT, DISPLAY_TE_PIN))
+      ;
+    while (GPIO_PIN_RESET == HAL_GPIO_ReadPin(DISPLAY_TE_PORT, DISPLAY_TE_PIN))
+      ;
+  }
+#endif
+}
+
+void display_set_compatible_settings(void) { display_panel_set_big_endian(); }
+
+static inline void set_window(const gl_bitblt_t* bb) {
+  display_panel_set_window(bb->dst_x, bb->dst_y, bb->dst_x + bb->width - 1,
+                           bb->dst_y + bb->height + 1);
+}
+
+void display_fill(const gl_bitblt_t* bb) {
```

Comment:
I absolutely agree; that's a good idea. I used display_fb.c and display_nofb.c—I don't particularly like these names, but at least they're consistent with the naming in the Rust code. I decided to use #ifdefs for conditional compilation of these files to avoid further complicating the build scripts.

see refactored code d4efd46b0996ffcde129811f1dcf747084df2281.

Reply:
I must admin, that you are right. see b0f7cc7ef4d29eefc090c41162e939fcab23a58b

Reply:
please also change the stm32f4 folder on u5 boards

Reply:
fixed here 4946d186f2583e0020042cac0d7128210d372e60

Reply:
fixed here 4946d186f2583e0020042cac0d7128210d372e60